api/client/trust.go

```diff
@@ -372,21 +372,30 @@ func (cli *DockerCli) trustedPush(repoInfo *registry.RepositoryInfo, tag string,
 
 	defer responseBody.Close()
 
-	targets := []target{}
+	// If it is a trusted push we would like to find the target entry which match the
+	// tag provided in the function and then do an AddTarget later.
+	target := &client.Target{}
 	handleTarget := func(aux *json.RawMessage) {
 		var pushResult distribution.PushResult
 		err := json.Unmarshal(*aux, &pushResult)
 		if err == nil && pushResult.Tag != "" && pushResult.Digest.Validate() == nil {
-			targets = append(targets, target{
-				reference: registry.ParseReference(pushResult.Tag),
-				digest:    pushResult.Digest,
-				size:      int64(pushResult.Size),
-			})
+			h, err := hex.DecodeString(pushResult.Digest.Hex())
+			if err != nil {
+				target = nil
+				return
+			}
+			target.Name = registry.ParseReference(pushResult.Tag).String()
+			target.Hashes = data.Hashes{string(pushResult.Digest.Algorithm()): h}
+			target.Length = int64(pushResult.Size)
```

Comment:
How about put the check for `tag` before the `DisplayJSONMessagesStream`, so that the `handleTarget` will never be called more than once. I mean:
```
        // We want trust signatures to always take an explicit tag,
        // otherwise it will act as an untrusted push.
        if tag == "" {
                if err = jsonmessage.DisplayJSONMessagesStream(responseBody, cli.out, cli.outFd, cli.isTerminalOut, nil); err != nil {
                        return err 
                }   
                fmt.Fprintf(cli.out, "No tag specified, skipping trust metadata push\n")
                return nil 
        }   
        if err = jsonmessage.DisplayJSONMessagesStream(responseBody, cli.out, cli.outFd, cli.isTerminalOut, handleTarget); err != nil {
                return err 
        }   

        if target == nil {
                fmt.Fprintf(cli.out, "No targets found, skipping trust metadata push\n")
                return nil 
        }
```